### ISSUE TITLE
Add pure Python implementation

### ISSUE THREAD

**Issue by maxbachmann** [maxbachmann/rapidfuzz#142]
Add a pure Python implementation which is reasonably efficient (more efficient than fuzzywuzzy)

**Comment by maxbachmann** [comment]
this is largely implemented in v2.1.0

**Comment by Rongronggg9** [comment]
> this is largely implemented in v2.1.0

I consider the new behavior (fallback silently) a breaking change... At least for me, I want the program immediately raise an exception if the cpp implementation is unavailable.
What's worse, it seems that there is no flag or warning indicating whether the fallback occurs or not, and there is no way to explicitly choose which implementation to use. That's really inconvenient.
Is it possible to at least add a flag or option? It is a fancy project and I really love it. But the new behavior prevents me from upgrading it to the latest version in my project.

**Comment by maxbachmann** [comment]
# Current fallback implementation

First let me explain the current fallback behavior:
1) when available it will always use the C++ implementation
2) wheels always include the C++ version and since wheels are available on most platforms most users will never use this slower implementation
3) even when no wheels are available it will first try to build the C++ version and only fallback to Python version if it is not possible to compile (In fact this is broken in v2.1.0, see #234 for the used mechanism, which will warn you when installing). The two main reasons why the compilation fails are:
  - the user does not has the necessary build tools installed (C++ compiler + cmake/ninja). Cmake and ninja are usually installed via the Python package, but currently this does not work properly on Arm
  - the user uses PyPy on a platform where scikit-build currently fails to build

So except for very few users everyone should always receive the C++ library. This change mostly gives users on more exotic platforms a reasonably fast implementation, which always works. In v2.1.1 you will receive a warning when installing the library and falling back to the Python version.
```
==========================================================================
WARNING: The C extension could not be compiled, speedups are not enabled.
Failure information, if any, is above.
Retrying the build without the C extension now.
==========================================================================

==========================================================================
WARNING: The C extension could not be compiled, speedups are not enabled.
Plain-Python build succeeded.
==========================================================================
```

#  Response to your concerns
I agree that it would be helpful for a library to be able to set which version they would like to receive. I think the best option would be to add something like `rapidfuzz.config.get/.set`. However I am unsure how this would affect the implementation. Maybe just raise an exception?
```
from rapidfuzz import config
config.allow_fallback = False -> exception Python fallback is used
```
would this be what your hoping for, or do you have a better idea how this could be implemented?

edit: in addition for build systems, where you REALLY want to build the C extension (e.g. for package managers) it is possible to set the environment variable `RAPIDFUZZ_BUILD_EXTENSION` before building the extension.

**Comment by Rongronggg9** [comment]
> First let me explain the current fallback behavior:

Thanks for your explanation. I exactly know that whether the fallback will occur has been determined during the installation of the package. What I am concerned about is just letting my program, at run time, know if the fallback occurs.

> I think the best option would be to add something like `rapidfuzz.config.get/.set`. However I am unsure how this would affect the implementation.

A better solution may be reading environment variables (maybe `RAPIDFUZZ_IMPLEMENTATION` = `cpp`/`python`?) so that the import graph of both your package and downstream package will not be too complex, I guess. And also makes force using pure Python implementation possible if a downstream really has such a demand.

**Comment by maxbachmann** [comment]
> A better solution may be reading environment variables (maybe RAPIDFUZZ_IMPLEMENTATION = cpp/python?) so that the import graph of both your package and downstream package will not be too complex, I guess. And also makes force using pure Python implementation possible if a downstream really has such a demand.

Sounds fine to me. Even though in cases where you control the system you might just want to set `RAPIDFUZZ_BUILD_EXTENSION` before installing the package. Note that a similar change will occur in https://github.com/maxbachmann/JaroWinkler, which is included as `rapidfuzz.distance.Jaro/JaroWinkler`.

**Comment by Rongronggg9** [comment]
> Even though in cases where you control the system you might just want to set `RAPIDFUZZ_BUILD_EXTENSION` before installing the package.

Exactly. However, if I distribute my program as a package on PyPI, it is impossible to control how to install its dependencies. But it can set an env var before importing dependencies :)

**Comment by maxbachmann** [comment]
I added this behavior to `v2.1.1`

**Comment by Alex-ley-scrub** [comment]
I also experienced this recently when moving from py311 to py312 - are there no py312 pre-built wheels @maxbachmann?
EDIT: I see there are wheels but for later versions https://github.com/rapidfuzz/RapidFuzz/releases/tag/v3.3.0

I only found that I was using the python fallback because one of my PR unit tests failed because the python fallback I am using is v2.15.1 gives different results to the cpp implementation - possible as you said here: https://github.com/rapidfuzz/RapidFuzz/issues/379#issuecomment-2074803294 it has an issue - which is causing a different result than we expected in our tests compared to the cpp results we used to get (and that we do get with py311 because it has pre-built wheels)

> There was a bug in the Python fallback implementation of fuzz.token_set_ratio that was fixed in version 3.6.0.

I really need to ensure that the cpp version is used for performance and so I'm now using this env var in my tests before I import rapidfuzz - and this catches it 💪  - so thank you:
```py
import os

os.environ["RAPIDFUZZ_IMPLEMENTATION"] = "cpp"

import rapidfuzz
```

Now I have to figure out why my cpp compiler isn't working when installing from source for py312 and/or upgrade to rapidfuzz>=3.3.1 and/or just use conda:
```
conda install -c conda-forge rapidfuzz=2.15.1
```

UPDATE: simply updating to 3.9.6 solved the `"RAPIDFUZZ_IMPLEMENTATION"` issue since it has wheels:
```
uv pip install -v --no-cache --only-binary rapidfuzz --reinstall-package rapidfuzz rapidfuzz>=3.3.1
```
but it also get's different results than `rapidfuzz==2.15.1` 🤔  so my tests still fail

UPDATE: I had left myself a code comment almost a year ago 🤣 - I should have listened to myself
```py
from rapidfuzz import process, fuzz
from rapidfuzz.utils import default_process  # might need to be used in rapidfuzz>=3.0

match_list = process.extract(  # Pulls out multiple matches
    input_text.lower(),
    examples.keys(),
    score_cutoff=95, 
    scorer=fuzz.WRatio,
    processor=default_process,  # to match results from v2 in v3
)
```

I had to add that arg `processor=default_process,` and then v3.9.6 gave me the same results as 2.15.1 🙌 

Hopefully my comment helps someone else who finds themself in a similar position resolve it quicker 🤣
